### ISSUE TITLE
thread 'tokio-runtime-worker' panicked 

### ISSUE THREAD

**Issue by kim201212** [sigp/lighthouse#2695]
## Description

Please provide a brief description of the issue.
```sh
Oct 08 15:23:59.588 INFO New block received                      hash: 0x21b083a9c70a56e5f26a03f492af08821d7d2fd09ccc582b8a39b388090b8cf1, slot: 2240218
Oct 08 15:24:05.001 INFO Synced                                  slot: 2240218, block: 0x21b0…8cf1, epoch: 70006, finalized_epoch: 70004, finalized_root: 0x0da1…85f5, peers: 50, service: slot_notifier
Oct 08 15:24:11.509 INFO New block received                      hash: 0x8b24e65b9bae465a932d1ad28e5443f5631deaca3386098817198904551fcef5, slot: 2240219
Oct 08 15:24:17.000 INFO Synced                                  slot: 2240219, block: 0x8b24…cef5, epoch: 70006, finalized_epoch: 70004, finalized_root: 0x0da1…85f5, peers: 50, service: slot_notifier
Oct 08 15:24:23.250 INFO New block received                      hash: 0x67e46e88d7ec7592561b5f60caada0032512a6d25c68d9212d2f8f930988b3d3, slot: 2240220
Oct 08 15:24:29.001 INFO Synced                                  slot: 2240220, block: 0x67e4…b3d3, epoch: 70006, finalized_epoch: 70004, finalized_root: 0x0da1…85f5, peers: 50, service: slot_notifier
Oct 08 15:24:36.120 INFO New block received                      hash: 0x692b430537bbf86471f3b8ac083b35abe8445742eb0a5d252b943f8b265be94c, slot: 2240221
Oct 08 15:24:41.000 INFO Synced                                  slot: 2240221, block: 0x692b…e94c, epoch: 70006, finalized_epoch: 70004, finalized_root: 0x0da1…85f5, peers: 51, service: slot_notifier
Oct 08 15:24:53.000 INFO Synced                                  slot: 2240222, block:    …  empty, epoch: 70006, finalized_epoch: 70004, finalized_root: 0x0da1…85f5, peers: 51, service: slot_notifier
Oct 08 15:24:59.702 INFO New block received                      hash: 0x6fc757e1e28625a2a6dcd177931be85956bd225861eea602a626a81739fc7b8c, slot: 2240223
Oct 08 15:25:05.000 INFO Synced                                  slot: 2240223, block: 0x6fc7…7b8c, epoch: 70006, finalized_epoch: 70004, finalized_root: 0x0da1…85f5, peers: 50, service: slot_notifier
Oct 08 15:25:14.242 INFO New block received                      hash: 0x272bef000b5d12be96de573c6629a7f38ab5cca6207bc9578a1b0c9d4bfa93fb, slot: 2240224
Oct 08 15:25:17.000 INFO Synced                                  slot: 2240224, block: 0x272b…93fb, epoch: 70007, finalized_epoch: 70005, finalized_root: 0x59a4…b065, peers: 49, service: slot_notifier
thread 'tokio-runtime-worker' panicked at 'internal error: entered unreachable code: The node was removed and shouldn't already exist', /usr/local/cargo/git/checkouts/discv5-7af694118de5b208/10247bb/src/kbucket/bucket.rs:451:21
stack backtrace:
   0: rust_begin_unwind
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/std/src/panicking.rs:493:5
   1: core::panicking::panic_fmt
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/core/src/panicking.rs:92:14
   2: discv5::kbucket::bucket::KBucket<TNodeId,TVal>::update_status
   3: discv5::kbucket::KBucketsTable<TNodeId,TVal>::update_node
   4: discv5::service::Service::connection_updated
   5: discv5::service::Service::handle_rpc_response
   6: discv5::service::Service::spawn::{{closure}}::{{closure}}
   7: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
   8: <futures_util::future::select::Select<A,B> as core::future::future::Future>::poll
   9: <futures_util::future::future::map::Map<Fut,F> as core::future::future::Future>::poll
  10: <futures_util::future::future::flatten::Flatten<Fut,<Fut as core::future::future::Future>::Output> as core::future::future::Future>::poll
  11: tokio::runtime::task::core::CoreStage<T>::poll
  12: tokio::runtime::task::harness::Harness<T,S>::poll
  13: std::thread::local::LocalKey<T>::with
  14: tokio::runtime::thread_pool::worker::Context::run_task
  15: tokio::runtime::thread_pool::worker::Context::run
  16: tokio::macros::scoped_tls::ScopedKey<T>::set
  17: tokio::runtime::thread_pool::worker::run
  18: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
  19: tokio::runtime::task::harness::Harness<T,S>::poll
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
Oct 08 15:25:22.251 CRIT Task panic. This is a bug!              advice: Please check above for a backtrace and notify the developers, message: <none>, task_name: discv5
Oct 08 15:25:22.252 INFO Internal shutdown received              reason: Panic (fatal error)
Oct 08 15:25:22.252 INFO Shutting down..                         reason: Failure("Panic (fatal error)")
Oct 08 15:25:22.255 INFO Saved DHT state                         service: network
Oct 08 15:25:22.255 INFO Network service shutdown                service: network
Oct 08 15:25:23.221 INFO Saved beacon chain to disk              service: beacon
Panic (fatal error)
```

## Version

```sh
$ docker images
REPOSITORY        TAG       IMAGE ID       CREATED       SIZE
sigp/lighthouse   v2.0.0    ce4de80f03ca   3 days ago    196MB

```

## Present Behaviour

BN stops any network activity and docker container stops.

## Expected Behaviour

BN performs network activity

## Steps to resolve

I restarted the stopped docker container.

**Comment by michaelsproul** [comment]
Thank you for reporting this!

This is a known bug in `discv5` that has been very difficult to track down. It existed in releases prior to v2.0.0 but we had never been able to reproduce it.

With the logging and backtraces added in v2.0.0 we finally have some more diagnostic information. We'll work on fixing it ASAP for a v2.x release.

In the meantime you can work around the issue by configuring the BN to auto-restart if it exits (something like `restart: on-failure` in Docker). We believe that the bug can only occur rarely, but please let us know if it occurs repeatedly

**Comment by AgeManning** [comment]
Thanks for raising this! 
These are exactly the logs we were chasing in order to narrow the issue down. 

It has been resolved in: https://github.com/sigp/discv5/pull/89

And I'll be propagating the updates into lighthouse in the next suite of network updates.

**Comment by SeaMonkey82** [comment]
Since the thread's still open, just in case additional logs are useful...

```
thread 'tokio-runtime-worker' panicked at 'internal error: entered unreachable code: The node was removed and shouldn't already exist', /home/seamonkey/.cargo/git/checkouts/discv5-7af694118de5b208/10247bb/src/kbucket/bucket.rs:451:21
stack backtrace:
   0: rust_begin_unwind
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/panicking.rs:515:5
   1: core::panicking::panic_fmt
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/panicking.rs:92:14
   2: discv5::kbucket::bucket::KBucket<TNodeId,TVal>::update_status
   3: discv5::kbucket::KBucketsTable<TNodeId,TVal>::insert_or_update
   4: discv5::service::Service::connection_updated
   5: discv5::service::Service::inject_session_established
   6: discv5::service::Service::spawn::{{closure}}::{{closure}}
   7: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
   8: <futures_util::future::select::Select<A,B> as core::future::future::Future>::poll
   9: <futures_util::future::future::map::Map<Fut,F> as core::future::future::Future>::poll
  10: <futures_util::future::future::flatten::Flatten<Fut,<Fut as core::future::future::Future>::Output> as core::future::future::Future>::poll
  11: tokio::runtime::task::core::CoreStage<T>::poll
  12: tokio::runtime::task::harness::Harness<T,S>::poll
  13: std::thread::local::LocalKey<T>::with
  14: tokio::runtime::thread_pool::worker::Context::run_task
  15: tokio::runtime::thread_pool::worker::Context::run
  16: tokio::macros::scoped_tls::ScopedKey<T>::set
  17: tokio::runtime::thread_pool::worker::run
  18: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
  19: tokio::runtime::task::harness::Harness<T,S>::poll
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
Oct 18 07:05:44.579 WARN Discovery query failed                  error: ChannelFailed("channel closed"), service: libp2p
Oct 18 07:05:44.581 CRIT Task panic. This is a bug!              advice: Please check above for a backtrace and notify the developers, message: <none>, task_name: discv5
Oct 18 07:05:44.582 INFO Internal shutdown received              reason: Panic (fatal error)
Oct 18 07:05:44.583 INFO Shutting down..                         reason: Failure("Panic (fatal error)")
Oct 18 07:05:44.586 INFO Saved DHT state                         service: network
Oct 18 07:05:44.586 INFO Network service shutdown                service: network
Oct 18 07:05:45.636 INFO Saved beacon chain to disk              service: beacon
Panic (fatal error)
```

**Comment by michaelsproul** [comment]
@SeaMonkey82 Did you get that panic in a version prior to v2.0.1? It should be fixed now in v2.0.1

I'll close this issue and we can re-open if it's still an issue in v2.0.1

**Comment by SeaMonkey82** [comment]
Got the error in 2.0.0.  Was just about to upgrade when I noticed my backup node had crashed.  I saw that it was fixed, but figured I'd go ahead and share the log just in case it might be useful - like if there were some variation in the details or root cause.
